### ISSUE TITLE
[Workflow] Questions and ideas

### ISSUE THREAD

**Issue by Padam87** [symfony/symfony#20774]
Hi,

I decided to give the workflow component a try, and I have a couple of questions and suggestions.

![graph](https://cloud.githubusercontent.com/assets/776488/20907033/e37f864a-bb4b-11e6-95ad-27a9fac6fa49.png)

I have started out with this workflow. I used to do these type of jobs with https://github.com/yohang/Finite, but the new component looks awesome :)

### Events

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L129

The marking store is updated only after the `enter` event has been fired. This is not convenient, since I can't flush the entity manager in the listener.
The `announce` event is for the next transition as I gather, so listening to that for a flush is not the way to go.
All in all, I have to add `$em->flush()` in my controller, outside my other logic related to that event to store the markings.

Could we add an  `entered` event to the mix ? :)

### Transition parameters

It would be useful to have the ability to pass parameters to transitions. In my past projects with https://github.com/yohang/Finite, parameters were a necessity sooner or later. I can't provide my exact use cases, but in the example above the `start_shipping` method could have a `tracking_number` parameter. I could of course set it in the controller, but then half of the transition logic will be in my listener, and half in the controller. eg

```php
if ($workflow->can($order, 'start_shipping')) {
    $order->setTrackingNumber('ASD123');
    $workflow->apply($order, 'start_shipping');
    $em->flush()
}
```

```php
public function onStartShipping(Event $event)
{
    /** @var Order $order */
    $order = $event->getSubject();

    $this->orderMailer->send(/* ... */) // The mail contains the tracking code
}
```

With finite, I used the `OptionsResolver` in my listeners to validate the passed options. In finite, we have also added a [built in resolver](https://github.com/yohang/Finite/pull/97/commits/8d3e07c2d5d2411373f3fa8cc291057c4f180795). (This may be a bit of overkill, especially because of the yaml configuration.)

I would like to submit a PR enabling this:
```php
if ($workflow->can($order, 'start_shipping')) {
    $workflow->apply($order, 'start_shipping', ['tracking_number' => 'ASD123']);
}
```
Would that be okay?

### One workflow, or multiple state machines?
The example could be implemented with 2 state machines. (`orderStatus`, `paymentStatus` and guards to validate)

In the documentation I read this: 

> It is also worth noting that a workflow does not commonly have cyclic path in the definition graph, but it is common for a state machine.

Would you recommend using a multi state workflow if there are cyclic paths (about 4-5) in a project?

Thanks!

**Comment by lyrixx** [comment]
## Event

> The announce event is for the next transition as I gather, so listening to that for a flush is not the way to go.
> All in all, I have to add $em->flush() in my controller, outside my other logic related to that event to store the markings.
> 
> Could we add an entered event to the mix ? :)

From my point of view, it's a best practice to flush in the controller and note in a listener. If you flush in the listener, your model could be corrupted.  For example, if you plan to send an email after a transition, nothing can guarantee you the email was send. But it's too late because you already save into the DB the new marking.

**Comment by lyrixx** [comment]
## Transition parameters

I'm not sure about this use case. You are mixing the concern here. I mean: You are using the event as a data bag and IMHO this is not the right place do it.

In your example, the tracking number should be a property of the Order. With the new implementation, the tracking number is volatile, and IMHO, it's not a good idea.

**Comment by Padam87** [comment]
Check the issues in finite... many people had problems with these there.

**Comment by lyrixx** [comment]
### One workflow, or multiple state machines?

In this example, and after a very quick read of it, I would use a workflow and not a state machine to let the Engine decide if a transition can be enabled or not.

**Comment by lyrixx** [comment]
@Padam87 Having only **big** issue for many question is really not easy :/

**Comment by Padam87** [comment]
Yep, that was a bad idea :)

I still think all of these are valid features, parameters come in handy in the real world, even if it does not make sense in an ideal word.

One time I had to pass parameters from an API call result, and on time from an rabbit RPC request.

**Comment by lyrixx** [comment]
@Padam87 Could we close this issue. I think I answered all your questions. Moreover there are related PR open or closed. Feel free to re-open new issue (but with only one topic ;) ) if needed.

**Comment by Padam87** [comment]
Yes, can be closed, thanks!
